### ISSUE TITLE
Current implementation of last saved height is prone to miss height.

### ISSUE THREAD

**Issue by sherpalden** [icon-project/centralized-relay#242]
```
func (r *Relayer) SaveChainsBlockHeight(ctx context.Context) {
    for _, chain := range r.chains {
        height, err := chain.Provider.QueryLatestHeight(ctx)
        if err != nil {
            r.log.Error("error occured when querying latest height", zap.Error(err))
            continue
        }
        if err := r.SaveBlockHeight(ctx, chain, height); err != nil {
            r.log.Error("error occured when saving block height", zap.Error(err))
            continue
        }
    }
}
```
Here, 
lets suppose latest height is 100 and we are processing at 95. Assume the relayer crashed after saving the latest height of 101.
Now, next time we run the relayer, we start polling from 101. But actually we need to do it from 95.

**Comment by ibrizsabin** [comment]
we need to save height only after processing is done

**Comment by debendraoli** [comment]
We use active listener using websocket. We don't save height everytime we process something. So saving height is periodic.
Detected events are immediately saved on disk, even it crashes it will go back to polling using the last saved height from disk.

Even subscription errors occurs we go back to polling and subscription at the same time.

**Comment by bcsainju** [comment]
I think this might cause an issue as the framework automatically saves height after x interval. This might work for chains with same implementation i.e. web sockets and fallback to rpc, this might not be true for all cases and might cause issue for other chains and the framework doesn't mandate the chains.

**Comment by debendraoli** [comment]
not having websocket, relying on poll would be matter of great concern in future.

This block interval saves does not mandate having websocket.

**Comment by debendraoli** [comment]
falling back to past height such as 5 mins ago would not cause an issue but missing might.

**Comment by ibrizsabin** [comment]
i think different people are talking about different things . it would be better if we had an architecture diagram of main loop. then it would be clearer for us all.

**Comment by debendraoli** [comment]
I can answer them in great details but the problem we're facing in the past and then is not missed hight but processing of messages that are detected.

We had have never missed a single message until now in the relayer, ever messages that are missed are in the relayer but not processed, which we've identified the issue.

I understand this is matter of concern but not right now.

**Comment by pragyanshrestha-ibriz** [comment]
@bcsainju  @debendraoli @sherpalden is this issue still relevant?

**Comment by bcsainju** [comment]
yes, this is still relevant but it's an edge case that might happen very rarely.

**Comment by debendraoli** [comment]
this is addressed here: #328
